toplevel/group/group.go

```diff
@@ -0,0 +1,317 @@
+package group
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"reflect"
+	"sort"
+
+	"github.com/app-sre/vault-manager/pkg/utils"
+	"github.com/app-sre/vault-manager/pkg/vault"
+	"github.com/app-sre/vault-manager/toplevel"
+	log "github.com/sirupsen/logrus"
+
+	"gopkg.in/yaml.v2"
+)
+
+type config struct{}
+
+var _ toplevel.Configuration = config{}
+
+type user struct {
+	Name  string `yaml:"org_username"`
+	Roles []role `yaml:"roles"`
+}
+
+type role struct {
+	Name        string           `yaml:"name"`
+	Permissions []oidcPermission `yaml:"oidc_permissions"`
+}
+
+type oidcPermission struct {
+	Name        string        `yaml:"name"`
+	Description string        `yaml:"description"`
+	Service     string        `yaml:"service"`
+	Policies    []vaultPolicy `yaml:"vault_policies"`
+}
+
+type vaultPolicy struct {
+	Name string `yaml:"name"`
+}
+
+type group struct {
+	Name      string
+	Id        string
+	Type      string
+	Metadata  map[string]interface{}
+	Policies  []string
+	EntityIds []string
+}
+
+func (g group) Key() string {
+	return g.Name
+}
+
+func (g group) KeyForType() string {
+	return g.Type
+}
+
+func (g group) KeyForDescription() string {
+	return fmt.Sprintf("%v", g.Metadata)
+}
+
+func (g group) Equals(i interface{}) bool {
+	group, ok := i.(group)
+	if !ok {
+		return false
+	}
+	return g.Name == group.Name &&
+		reflect.DeepEqual(g.Metadata, group.Metadata) &&
+		reflect.DeepEqual(g.Policies, group.Policies) &&
+		reflect.DeepEqual(g.EntityIds, group.EntityIds)
+}
+
+func (g group) CreateOrUpdate(action string) {
+	path := filepath.Join("identity", g.Type, "name", g.Name)
+	config := map[string]interface{}{
+		"member_entity_ids": g.EntityIds,
+		"policies":          g.Policies,
+		"metadata":          g.Metadata,
+	}
+	vault.WriteSecret(path, config)
+	log.WithField("path", path).WithField("type", g.Type).Info(
+		fmt.Sprintf("[Vault Identity] group successfully %s", action))
+}
+
+func (g group) Delete() {
+	path := filepath.Join("identity", g.Type, "name", g.Name)
+	vault.DeleteSecret(path)
+	log.WithField("path", path).WithField("type", g.Type).Info(
+		"[Vault Identity] group successfully deleted")
+}
+
+var _ vault.Item = group{}
+
+func init() {
+	toplevel.RegisterConfiguration("vault_groups", config{})
+}
+
+func (c config) Apply(entriesBytes []byte, dryRun bool, threadPoolSize int) {
```

Comment:
Considering how much it has to do, I love how clean and easy-to-follow this function is as written. 👍

Reply:
glad to hear it. It was in pretty rough shape for awhile :sweat_smile: